src/lib/auth/strategies/wpcom/oauth.js

```diff
@@ -0,0 +1,199 @@
+	 * overwrite the placeholder `_request`.
+	 *
+	 * @param {Object} token access token
+	 * @returns {function} authenticated promisfied wrapper for the 'wpcom-xhr-request' request
+	 * @private
+	 */
+	_createAuthenticatedRequest( token ) {
+		/**
+		 * It is basically a wrapper that promisifies and adds authentication to the request
+		 * method of 'wpcom-xhr-request'
+		 *
+		 * @param {Object} args request parameters
+		 * @returns {Promise} authenticated promisfied wrapper for the 'wpcom-xhr-request' request
+		 */
+		return ( args ) => {
+			return new Promise( ( resolve, reject ) => {
+				if ( ! token ) {
+					return reject( `Authentication ${ this.type } error: token is missing` );
+				}
+
+				const responseHandler = ( error, body ) => {
+					if ( error ) {
+						debug( `Request ${ args.method }:${ args.path } failed with : ${ error }` );
+						return reject( error );
+					}
+
+					debug( `${ args.method }:${ args.path } response ${ body }` );
+					return resolve( body );
+				};
+
+				wpComRequest( this._prepareParams( args ), responseHandler );
+			} );
+		};
+	}
+
+	/**
+	 * Placeholder methods that will be overwritten by a promisified authenticated request method
+	 * after the login is successfull.
+	 * @returns {Promise} rejects if the login failed and the method is not overwritten
+	 * @private
+	 */
+	_request() {
+		return Promise.reject( `Authentication ${ this.type } error: user not authenticated.` );
+	}
+
+	/**
+	 * Get the access token
+	 * @returns {string} The access token
+	 * @private
+	 */
+	_getToken() {
+		return this.token;
+	}
+
+	/**
+	 * Set the access token, used after a login or when we already have the token and we bypass the
+	 * oAuth login flow.
+	 * @param {string} token The access token
+	 */
+	setToken( token ) {
+		this.token = token;
+	}
+
+	/**
+	 * Login method for oAuth strategy. If the token was passed when instantiating this strategy
+	 * it will just overwrite the request method and return the promisifed token.
+	 *
+	 * If there is not access token saved, we follow the regular WPCOM oAuth flow in order to get
+	 * one, it is then saved and the request method is overwritten.
+	 *
+	 * It also overwrites the `_request` method with a promisified authenticated one
+	 *
+	 * @returns {Promise} the oauth access token
+	 */
+	login() {
+		const token = this._getToken();
+		if ( token ) {
+			debug( 'Token is present, customer is authenticated' );
+			// overwrite the request method to an authenticated promisified xhr request
+			this._request = this._createAuthenticatedRequest( token );
+
+			// bail if already authenticated
+			return Promise.resolve( token );
+		}
+
+		return new Promise( ( resolve, reject ) => {
+			debug( 'Get token, customer is not authenticated' );
+			oAuth.get( ( auth ) => {
+				if ( ! auth || ! auth.access_token ) {
+					reject( `Authentication ${ this.type } error: token is missing` );
```

Comment:
A `reject` won't stop further execution - might not be an issue, but wouldn't it be best to also explicitly `return` after this?

Reply:
Good catch, we should definitely return here.